commcare_connect/reports/views.py

```diff
@@ -106,14 +116,92 @@ def payment_strings(payment_data):
     }
 
 
-@login_required
-@user_passes_test(lambda user: user.is_superuser)
-@require_GET
-def delivery_stats_report(request):
-    table_data = []
-    quarters = _get_quarters_since_start()
-    for q in quarters:
-        data = _get_table_data_for_quarter(q)
-        table_data.append(data)
-    table = AdminReportTable(table_data)
-    return render(request, "reports/admin.html", context={"table": table})
+class SuperUserRequiredMixin(LoginRequiredMixin, UserPassesTestMixin):
+    def test_func(self):
+        return self.request.user.is_superuser
+
+
+class DeliveryReportFilters(django_filters.FilterSet):
+    delivery_type = django_filters.ChoiceFilter(method="filter_by_ignore")
+    year = django_filters.ChoiceFilter(method="filter_by_ignore")
+    quarter = django_filters.ChoiceFilter(
+        choices=[(1, "Q1"), (2, "Q2"), (3, "Q3"), (4, "Q4")], label="Quarter", method="filter_by_ignore"
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        current_year = datetime.now().year
+        year_choices = [(year, str(year)) for year in range(2023, current_year + 1)]
+        self.filters["year"] = django_filters.ChoiceFilter(
+            choices=year_choices, label="Year", method="filter_by_ignore"
+        )
+
+        delivery_types = DeliveryType.objects.values_list("slug", "name")
+        self.filters["delivery_type"] = django_filters.ChoiceFilter(choices=delivery_types, label="Delivery Type")
+
+    def filter_by_ignore(self, queryset, name, value):
+        return queryset
+
+    class Meta:
+        model = None
+        fields = ["delivery_type", "year", "quarter"]
+        unknown_field_behavior = django_filters.UnknownFieldBehavior.IGNORE
+
+
+class NonModelFilterView(FilterView):
+    def get_queryset(self):
+        # Doesn't matter which model it is here
+        return CompletedWork.objects.none()
+
+    @property
+    def object_list(self):
+        # Override this
+        return []
+
+    def get(self, request, *args, **kwargs):
+        filterset_class = self.get_filterset_class()
+        self.filterset = self.get_filterset(filterset_class)
+        context = self.get_context_data(filter=self.filterset, object_list=self.object_list)
+        return self.render_to_response(context)
+
+
+class DeliveryStatsReportView(tables.SingleTableMixin, SuperUserRequiredMixin, NonModelFilterView):
+    table_class = AdminReportTable
+    filterset_class = DeliveryReportFilters
+
+    def get_template_names(self):
+        if self.request.htmx:
```

Comment:
rather than add this new dependency and middleware, how would you feel about having the table just render asynchronously through a separate view call, like our other tables. That has the added benefit of not blocking the page render in the browser, which for one this slow, can be a much nicer experience, and hopefully requires less template duplication

Reply:
> render asynchronously through a separate view call, like our other tables.

May be I am not following this, this _is_ an async call. The table gets re-rendered when the filters are selected/modified.

I think the advantage of this is that we just need to define a single view that has all the logic in it. I agree this particular view is slow, but may be the priority to should be to improve that?